src/stapi_fastapi/models/product.py

```diff
@@ -54,33 +56,42 @@ class Product(BaseModel):
     _opportunity_properties: type[OpportunityProperties]
     _order_parameters: type[OrderParameters]
     _create_order: CreateOrder
-    _search_opportunities: SearchOpportunities
 
     def __init__(
         self,
         *args,
-        create_order: CreateOrder,
-        search_opportunities: SearchOpportunities,
         constraints: type[Constraints],
         opportunity_properties: type[OpportunityProperties],
         order_parameters: type[OrderParameters],
+        create_order: CreateOrder,
+        search_opportunities: SearchOpportunities | None = None,
+        search_opportunities_async: SearchOpportunitiesAsync | None = None,
+        get_opportunity_collection: GetOpportunityCollection | None = None,
         **kwargs,
     ) -> None:
         super().__init__(*args, **kwargs)
-        self._create_order = create_order
-        self._search_opportunities = search_opportunities
+
+        if bool(search_opportunities_async) != bool(get_opportunity_collection):
+            raise ValueError(
+                "Both the `search_opportunities_async` and `get_opportunity_collection` "
+                "arguments must be provided if either is provided"
+            )
+
         self._constraints = constraints
         self._opportunity_properties = opportunity_properties
         self._order_parameters = order_parameters
+        self._create_order = create_order
+        if search_opportunities is not None:
```

Comment:
Instead of conditionally setting it, I'd always set it, and then in supports_opportunity_search check for it's truthiness (not None). I don't think it's good practice to have Class-based objects that have different define fields in them -- I actually find it very odd the Python even has that behavior in objects, and I think it's probably a relic of the loosey-goosey dynamism that used to be more popular (probably still is popular, just not among those of us who follow a more type-checked approach).

Reply:
I think it was mypy that pushed me into that. Revised to always define the field and use properties with a callable-only return type (no None option).

Reply:
I'm interested in what the mypy warning was with that -- I'd expect that you have to set the retun type of the property to `| None`